hilti/toolchain/src/compiler/cxx/unit.cc

```diff
@@ -74,7 +74,11 @@ void Unit::_addModuleInitFunction() {
         add(extern_scope);
 
         cxx::Block register_;
-        register_.addStatement(fmt("%s = ::hilti::rt::Library::currentScope()", scope));
+
+        // Set the scope if it wasn't already set up.
+        register_.addIf(fmt("::hilti::rt::Library::isLoading() || %s == 0", scope),
```

Comment:
This is getting pretty complex/tricky for the generated code. How about a slightly different twist where we pass a  pointer to `__hlto_scope` into the runtime library and then centralize this logic there, like:
```
void hilti::rt::Library::setScope(uint64_t* scope) {
    if ( ::hilti::rt::Library::isLoading() || *scope == 0 )
        *scope = ::hilti::rt::Library::currentScope();
}
```

And we then probably don't even need `isLoading()` and `currentScope()` anymore, and can just inline those two inside `setScope()`, along with some generous comments explaining what's going on.

And then in unit.cc we simply generate:
```
    register_.addStatement(fmt("::hilti::rt::Library::setScope(&%s)", scope));
```

Reply:
Great idea. Let me change that to accept a mutable reference so we don't have to worry about null value.